README.md
<img src="https://banners.beyondco.de/Filament%20Address%20Picker.png?theme=light&packageManager=composer+require&packageName=erkurn%2Ffilament-address-picker&pattern=anchorsAway&style=style_2&description=Google+Map+API&md=1&showWatermark=0&fontSize=125px&images=globe&widths=300&heights=300" />

# Pick Address For Filament Field

[![Latest Version on Packagist](https://img.shields.io/packagist/v/erkurn/filament-address-picker.svg?style=flat-square)](https://packagist.org/packages/erkurn/filament-address-picker)
[![GitHub Tests Action Status](https://img.shields.io/github/workflow/status/erkurn/filament-address-picker/run-tests?label=tests)](https://github.com/erkurn/filament-address-picker/actions?query=workflow%3Arun-tests+branch%3Amain)
[![GitHub Code Style Action Status](https://img.shields.io/github/workflow/status/erkurn/filament-address-picker/Check%20&%20fix%20styling?label=code%20style)](https://github.com/erkurn/filament-address-picker/actions?query=workflow%3A"Check+%26+fix+styling"+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/erkurn/filament-address-picker.svg?style=flat-square)](https://packagist.org/packages/erkurn/filament-address-picker)

A Filament Field to enable pick location in google map and return geo coordinates.
For initial location you can use coordinate or address.

this packages require Google Map API and using this service : 
- Maps JavaScript API
- Places API

## Installation

You can install the package via composer:

```bash
composer require erkurn/filament-address-picker
```

## Usage

```php
<?php

class FilamentResource extends Resource
{
    public static function form(Form $form)
    {
        return $form->schema([
            AddressPicker::make('coordinate')
                ->setDefaultLocation([ // Set Default Location
                    'lat' =>    -6.914744,
                    'lng'  =>  107.609810
                ])
                ->mapControls([ // Map Controls
                    'mapTypeControl' => true,
                    'scaleControl' => true,
                    'streetViewControl' => false,
                    'rotateControl' => true,
                    'fullscreenControl' => true,
                    'searchBoxControl' => true
                ])
                ->minHeight(300) // Min Height In Pixels
                ->defaultZoom(16) // Default Zoom
                ->placeholder("Search Address") // Default Search
                ->afterStateUpdated(function ($state, $component, $set) {
                    $location = $component->getAddress(); // Get Details Location After Pick Location
                    
                    $location->getAdminLevels()->get(4)->getName(); // Post Code
                    $location->getStreetName(); // Street Name
                    $location->getStreetNumber(); // Street Number
                })
                ->placeholder("Search Address")
        ]);  
    }
}
```

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](https://github.com/erkurn/.github/blob/main/CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [Rezza Kurniawan](https://github.com/erkurn)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `filament-address-picker` will be documented in this file.

## Fix Errors - 2022-10-04

**Full Changelog**: https://github.com/erkurn/filament-address-picker/compare/v1.1.2...v1.1.3

## v1.1.2 - 2022-09-27

**Full Changelog**: https://github.com/erkurn/filament-address-picker/compare/v1.1.1...v1.1.2

## v1.1.1 - 2022-09-27

**Full Changelog**: https://github.com/erkurn/filament-address-picker/compare/v1.1.0...v1.1.1

## v1.1.0 - 2022-09-27

**Full Changelog**: https://github.com/erkurn/filament-address-picker/compare/v1.0.0...v1.1.0

## Initial Release - 2022-07-20

feat : add new field address picker

**Full Changelog**: https://github.com/erkurn/filament-address-picker/commits/v1.0.0


tests/Pest.php
<?php

use Erkurn\FilamentAddressPicker\Tests\TestCase;

uses(TestCase::class)->in(__DIR__);


tests/TestCase.php
<?php

namespace Erkurn\FilamentAddressPicker\Tests;

use Erkurn\FilamentAddressPicker\FilamentAddressPickerServiceProvider;
use Illuminate\Database\Eloquent\Factories\Factory;
use Orchestra\Testbench\TestCase as Orchestra;

class TestCase extends Orchestra
{
    protected function setUp(): void
    {
        parent::setUp();

        Factory::guessFactoryNamesUsing(
            fn (string $modelName) => 'Erkurn\\FilamentAddressPicker\\Database\\Factories\\'.class_basename($modelName).'Factory'
        );
    }

    protected function getPackageProviders($app)
    {
        return [
            FilamentAddressPickerServiceProvider::class,
        ];
    }

    public function getEnvironmentSetUp($app)
    {
        config()->set('database.default', 'testing');

        /*
        $migration = include __DIR__.'/../database/migrations/create_filament-address-picker_table.php.stub';
        $migration->up();
        */
    }
}


resources/js/components/address-picker.js
import { Loader } from '@googlemaps/js-api-loader';

export default (Alpine) => {
    Alpine.data('addressPickerFormComponent', ({
        state,
        api_key,
        zoom,
        controls,
        defaultLocation
    }) => {
        const loader = new Loader({
            apiKey: api_key,
            version: "weekly",
            libraries: ["places"],
        });

        return {
            state,
            api_key,
            markerLocation: {},
            init: function() {
                loader.loadCallback(e => {
                    if (e) {
                        console.log(e)
                    }

                    var valueLocation = null;
                    if (state instanceof Object) {
                        valueLocation = state.initialValue;
                    } else {
                        valueLocation = JSON.parse(state);
                    }

                    var center = {
                        lat: valueLocation?.lat || defaultLocation.lat,
                        lng: valueLocation?.lng || defaultLocation.lng
                    }

                    var map = new google.maps.Map(this.$refs.map_container, {
                        zoom: zoom,
                        center: center,
                        streetViewControl: false,
                        ...controls
                    });

                    var marker = new google.maps.Marker({
                        map
                    });

                    if (valueLocation?.lat && valueLocation?.lng) {
                        marker.setPosition(valueLocation);
                    }

                    map.addListener('click', (event) => {
                        this.markerLocation = event.latLng.toJSON();
                    });

                    if (controls.searchBoxControl) {
                        const input = this.$refs.map_search;
                        const searchBox = new google.maps.places.SearchBox(input);
                        map.controls[google.maps.ControlPosition.TOP_LEFT].push(input);
                        searchBox.addListener("places_changed", () => {
                            input.value = ''
                            this.markerLocation = searchBox.getPlaces()[0].geometry.location
                        })
                    }

                    this.$watch('markerLocation', () => {
                        let position = this.markerLocation;
                        this.state = position;
                        marker.setPosition(position);
                        map.panTo(position);
                    })
                })
            },
        }
    })
}


resources/views/components/address-picker.blade.php
<x-forms::field-wrapper
    :id="$getId()"
    :label="$getLabel()"
    :label-sr-only="$isLabelHidden()"
    :helper-text="$getHelperText()"
    :hint="$getHint()"
    :hint-icon="$getHintIcon()"
    :required="$isRequired()"
    :state-path="$getStatePath()"
>
    <div
        x-data="addressPickerFormComponent({
            state: $wire.entangle('{{ $getStatePath() }}'),
            api_key: '{{ $getApiKey() }}',
            zoom: {{$getDefaultZoom()}},
            controls: {{$getMapControls()}},
            defaultLocation: {{ $getDefaultLocation() }}
        })"
        {{ $attributes->merge($getExtraAttributes())->class(['relative filament-forms-address-picker-component']) }}
        wire:ignore
    >
        <div>
            @if($isSearchBoxControlEnabled())
                <input x-ref="map_search" type="text" placeholder="Search Location" style="margin: 10px 0 10px 0; width: 50%;" class="block transition duration-75 rounded-md shadow-sm focus:border-primary-500 focus:ring-1 focus:ring-inset focus:ring-primary-500 disabled:opacity-70 border-gray-300"/>
            @endif
            <div x-ref="map_container" style="min-height: {{ $getMinHeight()}}px; width: 100%; display: block;"></div>
        </div>
    </div>
</x-forms::field-wrapper>


src/Forms/Components/AddressPicker.php
<?php

namespace Erkurn\FilamentAddressPicker\Forms\Components;

use Filament\Forms\Components\Concerns\HasPlaceholder;
use Filament\Forms\Components\Field;
use Geocoder\Provider\GoogleMaps\GoogleMaps;
use Geocoder\Query\ReverseQuery;
use GuzzleHttp\Client;

class AddressPicker extends Field
{
    use HasPlaceholder;

    protected string $min_height;

    protected bool $debug = false;

    protected string $view = 'filament-address-picker::components.address-picker';

    public int $defaultZoom = 8;

    public array $controls = [
        'mapTypeControl' => true,
        'scaleControl' => true,
        'streetViewControl' => true,
        'rotateControl' => true,
        'fullscreenControl' => true,
        'searchBoxControl' => false,
    ];

    public array $defaultLocation = [
        'lat' => -6.914744,
        'lng' => 107.609810,
    ];

    public function getDefaultZoom(): int
    {
        return $this->defaultZoom;
    }

    public function getDefaultLocation(): string
    {
        return json_encode($this->defaultLocation, JSON_THROW_ON_ERROR);
    }

    public function setDefaultLocation(array $defaultLocation): static
    {
        $this->defaultLocation = $defaultLocation;

        return $this;
    }

    public function defaultZoom(int $defaultZoom): static
    {
        $this->defaultZoom = $defaultZoom;

        return $this;
    }

    public function getMapControls(): string
    {
        return json_encode($this->controls, JSON_THROW_ON_ERROR);
    }

    public function isSearchBoxControlEnabled(): bool
    {
        return $this->controls['searchBoxControl'];
    }

    public function mapControls(array $controls): static
    {
        $this->controls = array_merge($this->controls, $controls);

        return $this;
    }

    public function getApiKey()
    {
        return config('filament-address-picker.google_map_key');
    }

    public function getAddresses(): \Geocoder\Collection
    {
        $httpClient = new Client();
        $provider = new GoogleMaps($httpClient, null, $this->getApiKey());

        $geocoder = new \Geocoder\StatefulGeocoder($provider, 'en');

        return $geocoder->reverseQuery(ReverseQuery::fromCoordinates(
            data_get($this->getState(), 'lat'),
            data_get($this->getState(), 'lng')
        ));
    }

    public function getAddress(): \Geocoder\Location
    {
        return $this->getAddresses()->first();
    }

    public function getState()
    {
        $state = parent::getState();

        if (is_array($state)) {
            return $state;
        } else {
            if (count(explode(',', $state)) > 0 && ! empty($state)) {
                return [
                    'lat' => floatval(explode(',', $state)[0]),
                    'lng' => floatval(explode(',', $state)[1]),
                ];
            }

            try {
                return @json_decode($state, true, 512, JSON_THROW_ON_ERROR);
            } catch (\Exception $e) {
                return [
                    'lat' => 0,
                    'lng' => 0,
                ];
            }
        }
    }

    public function getMinHeight()
    {
        return $this->min_height ?? 500;
    }

    public function minHeight(int $height)
    {
        $this->min_height = $height;

        return $this;
    }

    public function getDebug()
    {
        return $this->debug;
    }

    public function showValue($value = true)
    {
        $this->debug = $value;

        return $this;
    }
}
